Enhance shape (#60335)

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63249 Enhance shape (#60335)**
* #62765 [ONNX] Suppport torch.dot and torch.nn.utils.spectral_norm (#62596)
* #62764 [ONNX] Update repeat_interleave for dynamic repeats (#59979)
* #62763 [ONNX] Fix an issue that optimizations might adjust graph inputs unexpectedly. (#61280)
* #62762 [ONNX] Fix controlflow shape inference with contrib op (#60707)
* #62760 [ONNX] Fix for batchnorm training op mode (#52758)
* #62759 [ONNX] Remove aten parameter (#61652)
* #62758 [ONNX] Add support for opset14 in PT-ONNX exporter (#59486)
* #62757 [ONNX] Support lstm_cell symbolic (#61476)

Enhanced shape inference by introducing typeReliableMap.

Co-authored-by: jiafatom <jiafa@microsoft.com>

Differential Revision: [D30375182](https://our.internmc.facebook.com/intern/diff/D30375182)

jiafatom committed:
Enhance shape (#60335)

Enhanced shape inference by introducing typeReliableMap.

Co-authored-by: jiafatom <jiafa@microsoft.com>

[ghstack-poisoned]

BowenBao committed:
Update on "Enhance shape (#60335)"


Enhanced shape inference by introducing typeReliableMap.

Co-authored-by: jiafatom <jiafa@microsoft.com>

[ghstack-poisoned]

BowenBao committed:
Update on "Enhance shape (#60335)"


Enhanced shape inference by introducing typeReliableMap.

Co-authored-by: jiafatom <jiafa@microsoft.com>

[ghstack-poisoned]

msaroufim commented:
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30375182).

BowenBao committed:
Update on "Enhance shape (#60335)"


Enhanced shape inference by introducing typeReliableMap.

Co-authored-by: jiafatom <jiafa@microsoft.com>

Differential Revision: [D30375182](https://our.internmc.facebook.com/intern/diff/D30375182)

[ghstack-poisoned]

ezyang commented:
This seems to have a lot of merge conflicts now:

```

Unmerged paths:
  (use "git restore --staged <file>..." to unstage)
  (use "git add <file>..." to mark resolution)
	both modified:   docs/source/onnx.rst
	both added:      test/onnx/expect/TestOperators.test_aten_embedding_1.expect
	both added:      test/onnx/expect/TestOperators.test_aten_embedding_2.expect
	both modified:   test/onnx/test_operators.py
	both modified:   torch/csrc/jit/passes/onnx.cpp
	both modified:   torch/csrc/jit/passes/onnx/shape_type_inference.cpp
	both modified:   torch/csrc/jit/passes/onnx/shape_type_inference.h
	both modified:   torch/onnx/utils.py
```

Do you mind rebasing?
